ethereum/contracts/BeefyLightClient.sol

```diff
@@ -260,7 +259,12 @@ contract BeefyLightClient {
         bytes32[] calldata mmrProofItems
     ) public {
         verifyCommitment(id, commitment, validatorProof);
-
+        verifyNewestMMRLeaf(
+            latestMMRLeaf,
+            mmrProofItems,
+            commitment.payload,
+            commitment.blockNumber
+        );
```

Comment:
This change implies that any commitment verification needs the full leaf, even if no validator set change is required. That increases gas consumption quite a bit I assume – do you plan to separate the two types of commitments in a later PR?

Reply:
Well we need to either:
 - always get the full leaf so that we can confirm the validator set id has not changed
 - check that the new change is within the same epoch so that we can assume there was no validator set change

adding a separate type would involve needing to implement the latter. we currently just track a MAXIMUM_BLOCK_GAP rather than tracking epochs/timestamps specifically, so it may not be that easy to check for an epoch change (eg: if 2 justifications come in a row at block 900 and block 902 but there was an epoch change/validator set change at 901) and would presumably add gas consumption to do.

there could be a cheaper/easier way, but i haven't thought about it too deeply yet and also given that the gas consumption addition for always sending the leaf is constant and not that significant, i'd say it's premature optimisation that we only need to worry about in future when we past the lower hanging fruit.

Reply:
Fully agree with you, my point here is that we already have that information in the commitment (from [here](https://github.com/Snowfork/polkadot-ethereum/blob/main/ethereum/contracts/BeefyLightClient.sol#L63-L67)): 

```Solidity
    struct Commitment {
        bytes32 payload;
        uint64 blockNumber;
        uint32 validatorSetId;
    }
```

So we only need the Beefy leaf once for the set change.

But agree, can optimize for that at a later point.

Reply:
oh right, ofc. yep makes sense :)